File: library/src/main/java/com/getstream/sdk/chat/adapter/AttachmentViewHolderMedia.java
package com.getstream.sdk.chat.adapter;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.bumptech.glide.Glide;
import com.getstream.sdk.chat.R;
import com.getstream.sdk.chat.enums.GiphyAction;
import com.getstream.sdk.chat.model.ModelType;
import com.getstream.sdk.chat.utils.Utils;
import com.getstream.sdk.chat.utils.roundedImageView.PorterShapeImageView;
import com.getstream.sdk.chat.view.MessageListView;
import com.getstream.sdk.chat.view.MessageListViewStyle;

import io.getstream.chat.android.client.models.Attachment;
import io.getstream.chat.android.client.models.Message;
import top.defaults.drawabletoolbox.DrawableBuilder;

public class AttachmentViewHolderMedia extends BaseAttachmentViewHolder {

    // Attachment
    private PorterShapeImageView iv_media_thumb;
    private TextView tv_media_title, tv_media_play, tv_media_des;
    private ImageView iv_command_logo;

    private ConstraintLayout cl_des, cl_action;
    private TextView tv_action_send, tv_action_shuffle, tv_action_cancel;
    private ProgressBar progressBar;

    private Context context;
    private Message message;
    private Attachment attachment;
    private MessageListItem.MessageItem messageListItem;
    private MessageListViewStyle style;
    private MessageListView.BubbleHelper bubbleHelper;
    private MessageListView.AttachmentClickListener clickListener;
    private MessageListView.MessageLongClickListener longClickListener;
    @NonNull
    private MessageListView.GiphySendListener giphySendListener;

    public AttachmentViewHolderMedia(
            int resId,
            @NonNull ViewGroup parent,
            @NonNull MessageListView.GiphySendListener giphySendListener
    ) {
        super(resId, parent);
        this.giphySendListener = giphySendListener;

        // Attach
        iv_media_thumb = itemView.findViewById(R.id.iv_media_thumb);
        tv_media_title = itemView.findViewById(R.id.tv_media_title);
        tv_media_play = itemView.findViewById(R.id.tv_media_play);
        tv_media_des = itemView.findViewById(R.id.tv_media_des);
        iv_command_logo = itemView.findViewById(R.id.iv_command_logo);
        cl_des = itemView.findViewById(R.id.cl_des);
        progressBar = itemView.findViewById(R.id.progressBar);
        // Giphy
        cl_action = itemView.findViewById(R.id.cl_action);
        tv_action_send = itemView.findViewById(R.id.tv_action_send);
        tv_action_shuffle = itemView.findViewById(R.id.tv_action_shuffle);
        tv_action_cancel = itemView.findViewById(R.id.tv_action_cancel);
    }

    @Override
    public void bind(@NonNull Context context,
                     @NonNull MessageListItem.MessageItem messageListItem,
                     @NonNull Message message,
                     @NonNull AttachmentListItem attachmentItem,
                     @NonNull MessageListViewStyle style,
                     @NonNull MessageListView.BubbleHelper bubbleHelper,
                     @Nullable MessageListView.AttachmentClickListener clickListener,
                     @Nullable MessageListView.MessageLongClickListener longClickListener) {
        this.context = context;
        this.messageListItem = messageListItem;
        this.message = message;
        this.attachment = attachmentItem.getAttachment();
        this.style = style;
        this.bubbleHelper = bubbleHelper;
        this.clickListener = clickListener;
        this.longClickListener = longClickListener;
        applyStyle();
        configMediaAttach();
        configAction();
        configClickListeners();
    }


    private void configImageThumbBackground() {
        Drawable background = bubbleHelper.getDrawableForAttachment(message,
                messageListItem.isMine(),
                messageListItem.getPositions(),
                attachment);
        iv_media_thumb.setShape(context, background);
    }

    private void configAction() {
        if (message.getType().equals(ModelType.message_ephemeral)
                && message.getCommand() != null
                && message.getCommand().equals(ModelType.attach_giphy)) {
            // set Background
            tv_action_send.setBackground(new DrawableBuilder()
                    .rectangle()
                    .rounded()
                    .strokeColor(Color.WHITE)
                    .strokeWidth(Utils.dpToPx(2))
                    .solidColor(context.getResources().getColor(R.color.stream_input_message_send_button))
                    .solidColorPressed(Color.LTGRAY)
                    .build());
            tv_action_shuffle.setBackground(new DrawableBuilder()
                    .rectangle()
                    .rounded()
                    .strokeColor(context.getResources().getColor(R.color.stream_message_stroke))
                    .strokeWidth(Utils.dpToPx(2))
                    .solidColor(Color.WHITE)
                    .solidColorPressed(Color.LTGRAY)
                    .build());
            tv_action_cancel.setBackground(new DrawableBuilder()
                    .rectangle()
                    .rounded()
                    .strokeColor(context.getResources().getColor(R.color.stream_message_stroke))
                    .strokeWidth(Utils.dpToPx(2))
                    .solidColor(Color.WHITE)
                    .solidColorPressed(Color.LTGRAY)
                    .build());

            cl_action.setVisibility(View.VISIBLE);

            tv_action_send.setOnClickListener(view -> {
                enableSendGiphyButtons(false);
                giphySendListener.onGiphySend(message, GiphyAction.SEND);
            });

            tv_action_shuffle.setOnClickListener((View v) -> {
                enableSendGiphyButtons(false);
                giphySendListener.onGiphySend(message, GiphyAction.SHUFFLE);
            });
            tv_action_cancel.setOnClickListener((View v) -> {
                giphySendListener.onGiphySend(message, GiphyAction.CANCEL);
            });
        } else {
            cl_action.setVisibility(View.GONE);
        }
    }

    private void configMediaAttach() {
        final String type = attachment.getType();
        configImageThumbBackground();
        final String imageUrl;
        if (attachment.getThumbUrl() != null && !attachment.getThumbUrl().isEmpty()) {
            imageUrl = attachment.getThumbUrl();
        } else {
            imageUrl = attachment.getImageUrl();
        }
        Glide.with(context)
                .asDrawable()
                .load(imageUrl)
                .placeholder(R.drawable.stream_placeholder)
                .into(iv_media_thumb);

        if (!message.getType().equals(ModelType.message_ephemeral))
            tv_media_title.setText(attachment.getTitle());
        tv_media_des.setText(attachment.getText());

        tv_media_title.setVisibility(!TextUtils.isEmpty(attachment.getTitle()) ? View.VISIBLE : View.GONE);
        tv_media_des.setVisibility(!TextUtils.isEmpty(attachment.getText()) ? View.VISIBLE : View.GONE);
        tv_media_play.setVisibility(type.equals(ModelType.attach_video) ? View.VISIBLE : View.GONE);
        iv_command_logo.setVisibility(type.equals(ModelType.attach_giphy) ? View.VISIBLE : View.GONE);

        if (tv_media_des.getVisibility() == View.VISIBLE || tv_media_title.getVisibility() == View.VISIBLE) {
            Drawable background = bubbleHelper.getDrawableForAttachmentDescription(messageListItem.getMessage(), messageListItem.isMine(), messageListItem.getPositions());
            cl_des.setBackground(background);
        }
    }

    private void applyStyle() {

        if (messageListItem.isMine()) {
            style.attachmentTitleTextMine.apply(tv_media_title);
            style.attachmentDescriptionTextMine.apply(tv_media_des);
        } else {
            style.attachmentTitleTextTheirs.apply(tv_media_title);
            style.attachmentDescriptionTextTheirs.apply(tv_media_des);
        }
    }

    private void enableSendGiphyButtons(boolean isEnable){
        progressBar.setVisibility(isEnable ? View.GONE : View.VISIBLE);
        tv_action_send.setEnabled(isEnable);
        tv_action_shuffle.setEnabled(isEnable);
        tv_action_cancel.setEnabled(isEnable);
    }

    private void configClickListeners(){
        iv_media_thumb.setOnClickListener(view -> {
            if (clickListener != null)
                clickListener.onAttachmentClick(message, attachment);
        });

        iv_media_thumb.setOnLongClickListener(view -> {
            if (longClickListener != null)
                longClickListener.onMessageLongClick(message);
            return true;
        });
    }
}

File: library/src/main/java/com/getstream/sdk/chat/adapter/viewholder/attachment/AttachmentViewHolderMedia.kt
package com.getstream.sdk.chat.adapter.viewholder.attachment

import android.graphics.Color
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import com.bumptech.glide.Glide
import com.getstream.sdk.chat.R
import com.getstream.sdk.chat.adapter.AttachmentListItem
import com.getstream.sdk.chat.adapter.MessageListItem.MessageItem
import com.getstream.sdk.chat.adapter.inflater
import com.getstream.sdk.chat.databinding.StreamItemAttachMediaBinding
import com.getstream.sdk.chat.enums.GiphyAction
import com.getstream.sdk.chat.model.ModelType
import com.getstream.sdk.chat.utils.Utils
import com.getstream.sdk.chat.view.MessageListView.AttachmentClickListener
import com.getstream.sdk.chat.view.MessageListView.BubbleHelper
import com.getstream.sdk.chat.view.MessageListView.GiphySendListener
import com.getstream.sdk.chat.view.MessageListView.MessageLongClickListener
import com.getstream.sdk.chat.view.MessageListViewStyle
import io.getstream.chat.android.client.models.Attachment
import io.getstream.chat.android.client.models.Message
import top.defaults.drawabletoolbox.DrawableBuilder

class AttachmentViewHolderMedia(
    parent: ViewGroup,
    private val giphySendListener: GiphySendListener,
    private val binding: StreamItemAttachMediaBinding =
        StreamItemAttachMediaBinding.inflate(parent.inflater, parent, false)
) : BaseAttachmentViewHolder(binding.root) {

    private lateinit var message: Message
    private lateinit var messageListItem: MessageItem
    private lateinit var style: MessageListViewStyle
    private lateinit var bubbleHelper: BubbleHelper

    private var clickListener: AttachmentClickListener? = null
    private var longClickListener: MessageLongClickListener? = null

    private lateinit var attachment: Attachment

    override fun bind(
        messageListItem: MessageItem,
        message: Message,
        attachmentListItem: AttachmentListItem,
        style: MessageListViewStyle,
        bubbleHelper: BubbleHelper,
        clickListener: AttachmentClickListener?,
        longClickListener: MessageLongClickListener?
    ) {
        this.messageListItem = messageListItem
        this.message = message
        this.style = style
        this.bubbleHelper = bubbleHelper
        this.clickListener = clickListener
        this.longClickListener = longClickListener

        attachment = attachmentListItem.attachment

        applyStyle()
        configMediaAttach()
        configActions()
        configClickListeners()
    }

    private fun applyStyle() {
        if (messageListItem.isMine) {
            style.attachmentTitleTextMine.apply(binding.tvMediaTitle)
            style.attachmentDescriptionTextMine.apply(binding.tvMediaDes)
        } else {
            style.attachmentTitleTextTheirs.apply(binding.tvMediaTitle)
            style.attachmentDescriptionTextTheirs.apply(binding.tvMediaDes)
        }
    }

    private fun configMediaAttach() {
        val type = attachment.type
        configImageThumbBackground()

        val thumbUrl = attachment.thumbUrl
        val imageUrl = if (thumbUrl.isNullOrEmpty()) attachment.imageUrl else thumbUrl

        Glide.with(context)
            .asDrawable()
            .load(imageUrl)
            .placeholder(R.drawable.stream_placeholder)
            .into(binding.ivMediaThumb)

        if (message.type != ModelType.message_ephemeral) {
            binding.tvMediaTitle.text = attachment.title
        }

        binding.tvMediaDes.text = attachment.text
        binding.tvMediaTitle.isVisible = !attachment.title.isNullOrEmpty()
        binding.tvMediaDes.isVisible = !attachment.text.isNullOrEmpty()
        binding.tvMediaPlay.isVisible = type == ModelType.attach_video
        binding.ivCommandLogo.isVisible = type == ModelType.attach_giphy

        if (binding.tvMediaDes.isVisible || binding.tvMediaTitle.isVisible) {
            val background = bubbleHelper.getDrawableForAttachmentDescription(
                messageListItem.message, messageListItem.isMine, messageListItem.positions
            )
            binding.clDes.background = background
        }
    }

    private fun configImageThumbBackground() {
        val background = bubbleHelper.getDrawableForAttachment(
            message,
            messageListItem.isMine,
            messageListItem.positions,
            attachment
        )
        binding.ivMediaThumb.setShape(context, background)
    }

    private fun configActions() {
        if (message.type == ModelType.message_ephemeral && message.command == ModelType.attach_giphy) {
            configGiphyAction()
        } else {
            binding.clAction.visibility = View.GONE
        }
    }

    private fun configGiphyAction() {
        binding.clAction.visibility = View.VISIBLE
        binding.tvActionSend.background = DrawableBuilder()
            .rectangle()
            .rounded()
            .strokeColor(Color.WHITE)
            .strokeWidth(Utils.dpToPx(2))
            .solidColor(ContextCompat.getColor(context, R.color.stream_input_message_send_button))
            .solidColorPressed(Color.LTGRAY)
            .build()
        binding.tvActionShuffle.background = DrawableBuilder()
            .rectangle()
            .rounded()
            .strokeColor(ContextCompat.getColor(context, R.color.stream_message_stroke))
            .strokeWidth(Utils.dpToPx(2))
            .solidColor(Color.WHITE)
            .solidColorPressed(Color.LTGRAY)
            .build()
        binding.tvActionCancel.background = DrawableBuilder()
            .rectangle()
            .rounded()
            .strokeColor(ContextCompat.getColor(context, R.color.stream_message_stroke))
            .strokeWidth(Utils.dpToPx(2))
            .solidColor(Color.WHITE)
            .solidColorPressed(Color.LTGRAY)
            .build()
        binding.tvActionSend.setOnClickListener {
            enableSendGiphyButtons(false)
            giphySendListener.onGiphySend(message, GiphyAction.SEND)
        }
        binding.tvActionShuffle.setOnClickListener {
            enableSendGiphyButtons(false)
            giphySendListener.onGiphySend(message, GiphyAction.SHUFFLE)
        }
        binding.tvActionCancel.setOnClickListener {
            giphySendListener.onGiphySend(
                message,
                GiphyAction.CANCEL
            )
        }
    }

    private fun enableSendGiphyButtons(isEnable: Boolean) {
        binding.progressBar.isVisible = !isEnable
        binding.tvActionSend.isEnabled = isEnable
        binding.tvActionShuffle.isEnabled = isEnable
        binding.tvActionCancel.isEnabled = isEnable
    }

    private fun configClickListeners() {
        binding.ivMediaThumb.setOnClickListener {
            clickListener?.onAttachmentClick(
                message,
                attachment
            )
        }
        binding.ivMediaThumb.setOnLongClickListener {
            longClickListener?.onMessageLongClick(message)
            true
        }
    }
}
